cmd/dockerd/daemon.go

```diff
@@ -198,6 +198,9 @@ func (cli *DaemonCli) start(opts *daemonOptions) (err error) {
 		cli.api.Accept(addr, ls...)
 	}
 
+	if err := registry.ValidateServiceOpts(cli.Config.ServiceOptions); err != nil {
+		return err
+	}
 	registryService := registry.NewService(cli.Config.ServiceOptions)
```

Comment:
How about adding a second (error) return value to `NewService()`? This would leave validation to the service and the errors when loading mirrors and registry, etc. wouldn't be ignored anymore (see a similar change in https://github.com/moby/moby/pull/34691/commits/bd3cf1ee56afd22321831396fe89ddf1813be602). Also, notice that the change in the current form is likely to break what @stevvooe suggest in https://github.com/moby/moby/pull/34319, namely to add private-registry-mirror semantics to `--registry-mirror`.

To me it seems like redundant checks, so I suggest to add a second error return value to `NewService()` and leave validation there.

Reply:
That's actually the best way to handle it. However, that change will require me updating many parts in the code.

Would you suggest to do that even though?

Reply:
@ripcurld0 Do you mean to apply the suggested changes or if I can take over this PR?

Reply:
Ah yes, that makes more sense. `newServiceConfig` should be the right place!

Reply:
@vrothberg 👍 
@AkihiroSuda @thaJeztah can you please give feedback.

Reply:
But we still need to pass error to daemon start, which implies adding an error return to `newService` (as it calls `newServiceConfig`. Doesn't it?

Reply:
Yes, actually we do otherwise there is no way to propagate that error.

Reply:
In a second thought we could create a new function called `NewServiceWithError` and use it in `daemon/dockerd.go`

Reply:
I think that a function that parses or validates user input should return an error, so I prefer to do that for `newService`. But that's just my 2 cents :-)